File: app/src/main/java/code/name/monkey/retromusic/model/AbsCustomPlaylist.java
package code.name.monkey.retromusic.model;

import android.content.Context;
import android.os.Parcel;

import java.util.ArrayList;

import androidx.annotation.NonNull;
import code.name.monkey.models.Song;
import io.reactivex.Observable;


public abstract class AbsCustomPlaylist extends Playlist {
    public AbsCustomPlaylist(int id, String name) {
        super(id, name);
    }

    public AbsCustomPlaylist() {
    }

    public AbsCustomPlaylist(Parcel in) {
        super(in);
    }

    @NonNull
    public abstract Observable<ArrayList<Song>> getSongs(Context context);
}

File: app/src/main/java/code/name/monkey/retromusic/model/AbsCustomPlaylist.kt
package code.name.monkey.retromusic.model

import code.name.monkey.retromusic.repository.LastAddedRepository
import code.name.monkey.retromusic.repository.SongRepository
import code.name.monkey.retromusic.repository.TopPlayedRepository
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject

abstract class AbsCustomPlaylist(
    id: Long,
    name: String
) : Playlist(id, name), KoinComponent {

    abstract fun songs(): List<Song>

    protected val songRepository by inject<SongRepository>()

    protected val topPlayedRepository by inject<TopPlayedRepository>()

    protected val lastAddedRepository by inject<LastAddedRepository>()
}